### PR TITLE
Added usecase and repository levels, search to home screen

### PR DESCRIPTION
**Pull request by AlexGolovach**

### CODE REVIEW

File: cocktails/src/main/java/com/renovavision/thecocktaildb/cocktails/details/CocktailDetailsViewModel.kt [hunk 1]
```diff
@@ -30,17 +34,17 @@ class CocktailDetailsViewModel @Inject constructor(
...
 
-    private fun loadCocktailInfo(id: Int) {
+    private fun loadCocktailInfo(drink: Drink) {
         getCocktailInfo.value = State(isLoading = true, showError = false)
 
         viewModelScope.launch(CoroutineExceptionHandler { _, _ ->
             getCocktailInfo.value = State(isLoading = false, showError = true)
         }) {
-            val cocktailInfo = cocktailsApi.loadCocktailInfoById(id)
+            val cocktailInfo = useCase.invoke(drink) as CocktailInfo
```
Comment (alexmprog) [inline cocktails/src/main/java/com/renovavision/thecocktaildb/cocktails/details/CocktailDetailsViewModel.kt]: @AlexGolovach Why you need cast here?

File: home/src/main/java/com/renovavision/thecocktaildb/home/HomeUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.renovavision.thecocktaildb.home
+
+import com.renovavision.thecocktaildb.network.DrinksByQuery.Drink
+import com.renovavision.thecocktaildb.network.data.repo.SearchRepo
+import javax.inject.Inject
+
+class HomeUseCase @Inject constructor(private val searchRepo: SearchRepo) {
+
+    private val drinksList = mutableListOf<Drink>()
```
Comment (alexmprog) [inline home/src/main/java/com/renovavision/thecocktaildb/home/HomeUseCase.kt]: @AlexGolovach Why do you need instance of `drinksList` here?

File: network/src/main/java/com/renovavision/thecocktaildb/network/data/Data.kt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package com.renovavision.thecocktaildb.network.data
+
+interface Data<T>{
```
Comment (alexmprog) [inline network/src/main/java/com/renovavision/thecocktaildb/network/data/Data.kt]: @AlexGolovach Rename interface instance to `Indexed`

File: network/src/main/java/com/renovavision/thecocktaildb/network/data/repo/CocktailsRepo.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import com.renovavision.thecocktaildb.network.data.Data
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class CocktailsRepo @Inject constructor(private val cocktailsApi: CocktailsApi) {
+
+    suspend fun loadData(data: Data<*>) = when (data) {
```
Comment (alexmprog) [inline network/src/main/java/com/renovavision/thecocktaildb/network/data/repo/CocktailsRepo.kt]: @AlexGolovach It's not a case for generic function. `CocktailsRepo` should provide 3 different methods here

File: network/src/main/java/com/renovavision/thecocktaildb/network/data/repo/CocktailsRepo.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        is Ingredient -> loadCocktailsListByIngredient(data.key)
+        is Category -> loadCocktailsListByCategory(data.key)
+        is Drink -> loadCocktailDetails(data.key)
+        else -> throw IllegalArgumentException()
+    }
+
+    private suspend fun loadCocktailsListByIngredient(ingredient: String) =
+        withContext(Dispatchers.Default) {
```
Comment (alexmprog) [inline network/src/main/java/com/renovavision/thecocktaildb/network/data/repo/CocktailsRepo.kt]: @AlexGolovach Why do you need `withContext` coroutine here?

**Review by alexmprog — CHANGES_REQUESTED** [review]

**Comment by alexmprog** [comment]
@AlexGolovach All `usecase` classes should be called in right way ...`GetCocktailsByCategory`, `SearchCocktails` and etc.

File: cocktails/src/main/java/com/renovavision/thecocktaildb/cocktails/details/CocktailDetailsViewModel.kt (hunk 1 above)
Comment (AlexGolovach) [reply cocktails/src/main/java/com/renovavision/thecocktaildb/cocktails/details/CocktailDetailsViewModel.kt]: Fixed

File: home/src/main/java/com/renovavision/thecocktaildb/home/HomeUseCase.kt (hunk 2 above)
Comment (AlexGolovach) [reply home/src/main/java/com/renovavision/thecocktaildb/home/HomeUseCase.kt]: Fixed

File: network/src/main/java/com/renovavision/thecocktaildb/network/data/Data.kt (hunk 3 above)
Comment (AlexGolovach) [reply network/src/main/java/com/renovavision/thecocktaildb/network/data/Data.kt]: Fixed

File: network/src/main/java/com/renovavision/thecocktaildb/network/data/repo/CocktailsRepo.kt (hunk 4 above)
Comment (AlexGolovach) [reply network/src/main/java/com/renovavision/thecocktaildb/network/data/repo/CocktailsRepo.kt]: Fixed

File: network/src/main/java/com/renovavision/thecocktaildb/network/data/repo/CocktailsRepo.kt (hunk 5 above)
Comment (AlexGolovach) [reply network/src/main/java/com/renovavision/thecocktaildb/network/data/repo/CocktailsRepo.kt]: Fixed

**Pull request merged**